### ISSUE TITLE
I would like to use Bulk API 2.0

### ISSUE THREAD

**Issue by geospatialqueryapi** [forcedotcom/SalesforceMobileSDK-iOS#3461]
I would like to use Bulk API 2.0 from
my native iOS App.

How is the best way to accomplish that?

Thank you in advance!

**Comment by wmathurin** [comment]
We don't have wrapper methods for bulk api 2.0 calls but one can build SFRestRequest for arbitrary end point (see https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/dev/libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.h). You don't have to worry about authentication. The SDK's sendRequest method will take care of adding the access token header, and if if has expired, a refresh is automatically done, and the request is "replayed".

**Comment by geospatialqueryapi** [comment]
Thank you very much!
I believe it should be like this:

let dataString = "{
      "object" : "<myObject>__c",
      "externalIdFieldName" : "<myID>__c",
      "contentType" : "CSV",
      "operation" : "upsert"
  }"

if let data = dataString.data(using: .utf8) {
    
    let request = RestRequest(method: .POST, serviceHostType: .instance, 
        path: "/services/data/v52.0/jobs/ingest", queryParams: nil) 
    request.setCustomRequestBodyData(data, contentType: "application/json")
    
    RestClient.shared.send(request: request { [weak self] (result) in
        switch result {
            case .success(let response):
                // do something with response
            case .failure(let error):
                SalesforceLogger.d(RootViewController.self, message:"Error invoking: \(request) , \(error)")
        } 
    }
}

**Comment by geospatialqueryapi** [comment]
I have alternative way of doing it. I'm using Golang and Gomobile for 
embedding a binary in iOS App.

The only information I need is the sessionID.

	Bearer := "Bearer " + sessionId
	req.Header.Add("Authorization", Bearer)

Also the sessions should be refreshed.
How to get the SessionID from the Swift code?

**Comment by wmathurin** [comment]
You can get the access token (session id) from the current user.
```swift
let accessToken = UserAccountManager.shared.currentUserAccount?.credentials.accessToken
```

**Comment by geospatialqueryapi** [comment]
Will this call return updated accessToken if it expires?

**Comment by geospatialqueryapi** [comment]
There is one other SDK behavior which is bothering me. 
When the login screen appears for 1st time, I'm switching 
to Notes to copy my username, the App is going in background
and when the login screen appears again it clears all the fields.

**Comment by wmathurin** [comment]
If you use our RestClient to send requests then we intercept the 401 returned by the server when the token expires, go to the oauth end point to refresh the token and then resend the original request with the new token. All of it being invisible to the application. If you send the requests through other means, then you have to do all of that yourself.

**Comment by geospatialqueryapi** [comment]
I'll definitely use the RestClient then.
Would like to have AccountsListView running on a tab.
I would like to initiate the Authentication process when the users clicks 
on its tab, rather from SceneDelegate. SceneDelegate is not recommended 
anyway for new Swift UI Apps.

How to accomplish Auth then?

**Comment by wmathurin** [comment]
We have no plans of adding out of the box support for bulk api. Instead we are adding rest wrappers (and response classes) for using sobject collection APIs which allows for CRUD operations on a list of objects.
They will be available in our upcoming release: 10.1.
For a sneak peak, see: https://github.com/forcedotcom/SalesforceMobileSDK-iOS/pull/3507
